Source/EasyNetQ/Producer/PublishConfirmationWaiter.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EasyNetQ.Internals;
+
+namespace EasyNetQ.Producer
+{
+    public class PublishConfirmationWaiter : IPublishConfirmationWaiter
+    {
+        private readonly ulong deliveryTag;
+        private readonly Task confirmation;
+        private readonly CancellationToken cancellation;
+        private readonly Action action;
+
+        public PublishConfirmationWaiter(ulong deliveryTag, Task confirmation, CancellationToken cancellation, Action action)
+        {
+            this.deliveryTag = deliveryTag;
+            this.confirmation = confirmation;
+            this.cancellation = cancellation;
+            this.action = action;
+        }
+
+        public void Wait(TimeSpan timeout)
+        {
+            try
+            {
+                if (confirmation.Wait((int)timeout.TotalMilliseconds, cancellation))
+                {
+                    return;
+                }
+
+                throw new TimeoutException(string.Format("Publisher confirms timed out after {0} seconds waiting for ACK or NACK from sequence number {1}", (int)timeout.TotalSeconds, deliveryTag));
+            }
+            catch (AggregateException exception)
+            {
+                throw exception.InnerException;
+            }
+            finally
+            {
+                action();
+            }
+        }
+
+        public async Task WaitAsync(TimeSpan timeout)
+        {
+            try
+            {
+                using (var timeoutCancellation = new CancellationTokenSource())
+                {
+                    using (var compositeCancellation = CancellationTokenSource.CreateLinkedTokenSource(timeoutCancellation.Token, cancellation))
+                    {
+                        var timeoutTask = TaskHelpers.Delay(timeout, compositeCancellation.Token);
+                        if (timeoutTask == await TaskHelpers.WhenAny(confirmation, timeoutTask).ConfigureAwait(false))
+                        {
+                            throw new TimeoutException(string.Format("Publisher confirms timed out after {0} seconds waiting for ACK or NACK from sequence number {1}", (int)timeout.TotalSeconds, deliveryTag));
+                        }
+                        timeoutCancellation.Cancel();
```

Comment:
Here is an answer:

```csharp
public static Task Delay(int dueTime, CancellationToken cancellationToken)
    {
      if (dueTime < -1)
        throw new ArgumentOutOfRangeException("dueTime", "The timeout must be non-negative or -1, and it must be less than or equal to Int32.MaxValue.");
      Contract.EndContractBlock();
      if (cancellationToken.IsCancellationRequested)
        return new Task((Action) (() => {}), cancellationToken);
      if (dueTime == 0)
        return TaskEx.s_preCompletedTask;
      TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
      CancellationTokenRegistration ctr = new CancellationTokenRegistration();
      Timer timer = (Timer) null;
      timer = new Timer((TimerCallback) (state =>
      {
        ctr.Dispose();
        timer.Dispose();
        tcs.TrySetResult(true);
        TimerManager.Remove(timer);
      }), (object) null, -1, -1);
      TimerManager.Add(timer);
      if (cancellationToken.CanBeCanceled)
        ctr = cancellationToken.Register((Action) (() =>
        {
          timer.Dispose();
          tcs.TrySetCanceled();
          TimerManager.Remove(timer);
        }));
      timer.Change(dueTime, -1);
      return (Task) tcs.Task;
    }
```

So, with timeoutCancellation we could release timer as soon as possible. In case of highload and very long timeouts we could have a lot of unreleased timers.